internals/cli/masker/writer.go

```diff
@@ -1,225 +1,171 @@
-// - When the incoming bytes do not end in a partial secret, they are directly passed to the output channel.
-// - When the incoming bytes do end in a partial secret, all bytes up to the partial secret are passed to the
-//   output channel. Passing on the partial secrets bytes is delayed until:
-//     - New bytes come in that finish the secret, in which case the bytes are passed to the output channel and marked as masked.
-//     - New bytes come in that do not finish the secret, in which case the bytes are passed to the output channel and not marked as masked.
-//     - The output timeout channel receives a signal, in which case the bytes are passed to the output channel and not marked as masked.
-func (mw *MaskedWriter) process() {
+func (m *Masker) Run(ctx context.Context) {
+	m.flushAllChan = make(chan struct{})
 	for {
 		select {
-		case <-mw.outputTimeoutCh:
-			// Only flush if there is still nothing send to the output channel.
-			if len(mw.outputCh) == 0 {
-				for _, matcher := range mw.matchers {
-					matcher.Reset()
-				}
-				mw.flushBuffer()
-			}
-		case p := <-mw.incomingBytesCh:
-			for _, b := range p {
-				matchInProgress := false
-				mw.buf = append(mw.buf, maskByte{byte: b})
-
-				for _, matcher := range mw.matchers {
-					maskLen := matcher.Read(b)
-					for i := 0; i < maskLen; i++ {
-						mw.buf[len(mw.buf)-1-i].masked = true
-					}
-					matchInProgress = matchInProgress || matcher.InProgress()
-				}
-
-				if !matchInProgress {
-					mw.flushBuffer()
-				}
+		case <-m.flushAllChan:
+			m.flushAll()
+			return
+		default:
+		}
+
+		if len(m.triggers) == 0 {
+			select {
+			case <-time.After(m.BufferDelay / 2):
+				continue
+			case <-m.flushAllChan:
+				m.flushAll()
+				return
 			}
 		}
-	}
-}
 
-func (mw *MaskedWriter) flushBuffer() {
-	tmp := make([]maskByte, len(mw.buf))
-	copy(tmp, mw.buf)
-	mw.outputCh <- tmp
-	mw.buf = mw.buf[:0]
-}
+		m.lock.Lock()
+		trigger := m.triggers[0]
+		m.lock.Unlock()
 
-// Run writes any processed data from the output channel to the underlying io.Writer.
-// If no new data is received on the output channel for timeout, the output buffer is forced flushed
-// and all ongoing matches are reset.
-//
-// This should be run in a separate goroutine.
-func (mw *MaskedWriter) Run() {
-	go mw.process()
-	masking := false
-	for {
 		select {
-		case output := <-mw.outputCh:
-			for _, b := range output {
-				var err error
-				if b.masked {
-					if !masking {
-						_, err = mw.w.Write([]byte(mw.maskString))
-						if err != nil {
-							mw.errCh <- err
-							return
-						}
-					}
-					masking = true
-				} else {
-					_, err = mw.w.Write([]byte{b.byte})
-					if err != nil {
-						mw.errCh <- err
-						return
-					}
-					masking = false
-				}
-			}
-			mw.wg.Add(-len(output))
-		case <-time.After(mw.timeout):
-			// send to the timeout channel if not already done so.
-			select {
-			case mw.outputTimeoutCh <- struct{}{}:
-			default:
+		case <-trigger.timer.C:
```

Comment:
Maybe we could check whether the next trigger is also done and combine them? That way, we don't check for matches multiple times when not necessary.

Reply:
This depends on how we want to solve the secret matching logic. If we make that more like the previous implementation (with a stateful matcher), the number of times `flushN()` is called, does not affect the performance.